### PR TITLE
Feature/incremental predicates

### LINKED ISSUE #160: support user incremental predicates on merge
### Describe the feature

Users should be able to supply incremental models with arbitrary predicates during the merge step to prevent unnecessary table scan for massive data sets. 

### Describe alternatives you've considered
None

### Additional context
See: 
- [Core Issue 5680](https://github.com/dbt-labs/dbt-core/issues/5680)
- [Core Issue 3293](https://github.com/dbt-labs/dbt-core/issues/3293)
- [Core PR 5702](https://github.com/dbt-labs/dbt-core/pull/5702)

### Who will this benefit?
users with extremely large datasets who want to limit the scan of their merge operations on incremental models!

### Are you interested in contributing this feature?
Yes

### PR DESCRIPTION
**Pull request by dave-connors-3**
resolves #160

### Description

Updates the incremental materialization to include user supplied `incremental_predicates` and passes that to the `dbt_spark_get_incremental_sql` macro. 

The changes to `dbt_spark_get_incremental_sql` are updated in [dbt-spark PR #436](https://github.com/dbt-labs/dbt-spark/pull/436)

### Checklist

- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-databricks next" section.

### CODE REVIEW

File: dbt/include/databricks/macros/materializations/incremental/incremental.sql
Comment (ueshin) [inline dbt/include/databricks/macros/materializations/incremental/incremental.sql:5]: Could you revert this? `default` for `incremental_strategy` won't work. It should be `or 'merge'`.

File: dbt/include/databricks/macros/materializations/incremental/incremental.sql
Comment (ueshin) [inline dbt/include/databricks/macros/materializations/incremental/incremental.sql:6]: nit: Shall we move this line to the next line of `{%- set incremental_strategy = ... -%}`?

**Review by ueshin — COMMENTED** [review]
Otherwise, LGTM.
I'll test with this change once the upstreams are merged.

File: dbt/include/databricks/macros/materializations/incremental/incremental.sql
Comment (dave-connors-3) [reply dbt/include/databricks/macros/materializations/incremental/incremental.sql:6]: @ueshin just to confirm, you'd prefer the predicates line on line 10?

Comment (ueshin) [reply dbt/include/databricks/macros/materializations/incremental/incremental.sql:6]: Yes, I'd prefer to put it in the same block as the other configs, not the temporary ones.

File: dev-requirements.txt
Comment (ueshin) [inline dev-requirements.txt:8]: Are these changes still necessary? If so, we need to wait for the changes in `dbt-core` and `dbt-spark`.

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:47]: Could you add a blank line at the end of file?

File: tests/integration/incremental_strategies/test_incremental_strategies.py
Comment (ueshin) [inline tests/integration/incremental_strategies/test_incremental_strategies.py:71]: Something went wrong. Could you fix it?

File: dev-requirements.txt
Comment (dave-connors-3) [reply dev-requirements.txt:8]: for now, yes! hoping to merge the `dbt-core` and `dbt-spark` PRs shortly!

File: tests/integration/incremental_strategies/test_incremental_strategies.py
Comment (dave-connors-3) [reply tests/integration/incremental_strategies/test_incremental_strategies.py:71]: embarrassing! taking a look now

**Comment by dave-connors-3** [comment]
@ueshin thank you for the review! I believe i've addressed the outstanding issues!

**Comment by ueshin** [comment]
Hi @dave-connors-3, now that the PRs for `dbt-core` and `dbt-spark` were merged.
Could you revert `dev-requirements.txt` and fix the linter errors?

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:38]: I don't think we need these as the default strategy is `merge` and the default file format is `delta` in dbt-databricks.

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:58]: ditto.

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:5]: Do we need to re-define this?

Comment (dave-connors-3) [reply tests/functional/adapter/test_incremental_predicates.py:5]: i believe I needed to redefine this in my [dbt-spark PR](https://github.com/dbt-labs/dbt-spark/pull/436/files#diff-476253c5fe28ecc18fa10e0819037b5ffe7e877d781aa83199c3d42961a3615e), and I assumed I needed to do the same here. 

i did just run the SQL in databricks without the cast function and it seemed to execute. I can remove this if we don't think this is necessary!

**Comment by ueshin** [comment]
Why was `tests/integration/incremental_strategies/seeds/expected_exclude_upsert.csv` removed?

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [reply tests/functional/adapter/test_incremental_predicates.py:5]: Shall we just rename it to `models__databricks_incremental_predicates_sql`, then?

**Comment by dave-connors-3** [comment]
> Why was `tests/integration/incremental_strategies/seeds/expected_exclude_upsert.csv` removed? Could you add it back?

Totally by mistake! restored the file

**Review by ueshin — APPROVED** [review]
LGTM, pending linter.
@dave-connors-3 Could you run `black` again?

**Comment by dave-connors-3** [comment]
@ueshin i think black is passing, but flake8 line length checks are failing

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:39]:
```suggestion
        return {
            "delete_insert_incremental_predicates.sql": (
                models__databricks_incremental_predicates_sql
            )
        }
```

File: tests/functional/adapter/test_incremental_predicates.py
Comment (ueshin) [inline tests/functional/adapter/test_incremental_predicates.py:51]:
```suggestion
        return {
            "delete_insert_incremental_predicates.sql": (
                models__databricks_incremental_predicates_sql
            )
        }
```

**Comment by ueshin** [comment]
@dave-connors-3 Yeah, sometimes `black` and `flake8` conflict..

**Comment by ueshin** [comment]
Thanks! merging.

**Pull request merged**
